### PR TITLE
Add scala3 support for the java module

### PR DESCRIPTION
**Pull request by regadas**

### CODE REVIEW

**Comment by regadas** [comment]
@RustedBones I think this can work?

File: java/src/main/scala/com/spotify/featran/java/SerializableFunction.scala [hunk 1]
```diff
@@ -15,10 +15,8 @@
...
-package com.spotify.featran.java;
+package com.spotify.featran.java
 
-import java.io.Serializable;
+import java.util.function.{Function => JFunction}
 
-public interface SerializableFunction<InputT, OutputT> extends Serializable {
-  OutputT apply(InputT input);
-}
+trait SerializableFunction[I, O] extends JFunction[I, O] with Serializable
```
Comment (RustedBones) [inline java/src/main/scala/com/spotify/featran/java/SerializableFunction.scala:22]: binary compatibility relies on compilation to translate to java interface bytecode.

**Review by RustedBones — COMMENTED** [review]
I think the original goal was to expose java only API. Here we introduce back some scala code

File: java/src/main/scala/com/spotify/featran/java/JavaOps.scala [hunk 2]
```diff
@@ -20,25 +20,25 @@ package com.spotify.featran.java
...
   def requiredFn[I, O](f: SerializableFunction[I, O]): I => O =
-    (input: I) => f(input)
+    f.asScala
 
   def optionalFn[I, O](f: SerializableFunction[I, Optional[O]]): I => Option[O] =
-    (input: I) => {
-      val o = f(input)
-      if (o.isPresent) Some(o.get()) else None
-    }
+    f.asScala.andThen(_.asScala)
```
[the remark this answers is not in the document]
Comment (regadas) [reply java/src/main/scala/com/spotify/featran/java/JavaOps.scala:41]: Yeah, that's true … here, we might end up with forward compat issues. Let's revert this, then.

**Review by regadas — COMMENTED** [review]
Closing in favor of #679

**Pull request closed without merging**